### DISCUSSION TITLE
issue: useFieldArray incorrectly treats value of 0 as empty value

### DISCUSSION THREAD

**Discussion by deshiknaves** [react-hook-form/react-hook-form#12018] [Q&A]
### Version Number

7.51.5

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/react-hook-form-array-zvvwvq?file=%2Fsrc%2FApp.tsx%3A23%2C12&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxfgtlc500063p6h3kazjlpx%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxfgtlc500023p6h174pug6j%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxfgtlc500033p6hzso2nh3v%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxfgtlc500053p6hpxvski26%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxfgtlc500023p6h174pug6j%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxfgtlc500013p6hdviz1r6u%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxfgyvan00023p6gcpna98ar%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A23%252C%2522startColumn%2522%253A12%252C%2522endLineNumber%2522%253A23%252C%2522endColumn%2522%253A12%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxfgtlc500023p6h174pug6j%2522%252C%2522activeTabId%2522%253A%2522clxfgyvan00023p6gcpna98ar%2522%257D%252C%2522clxfgtlc500053p6hpxvski26%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxfgtlc500043p6h41oga3mj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clxfgtlc500053p6hpxvski26%2522%252C%2522activeTabId%2522%253A%2522clxfgtlc500043p6h41oga3mj%2522%257D%252C%2522clxfgtlc500033p6hzso2nh3v%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clxfgtlc500033p6hzso2nh3v%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D

## First
1. Click `Add`
2. Type `0` in the field
3. Click `Add`

## Second
1. Click `Add`
2. Type `1` in the new field
3. Click `Add`
4. Type `2` in the new field
5. Click `Add`
6. Type `3` in the new field
7. Focus field with `2`
8. Highlight the text
9. Hit `delete`
10. Click `Add`

### Expected behaviour

It should add a new field in the first scenario.

The problem is that `_getFieldArray` uses the `utils/compact` method which uses `.filter(Boolean)`. For a number field `0` is a filled value and so we should be able to add another field.

To me it feels like it should be an option if the values should be compacted or not. For example in the second example, I would expect it to add another field at the end.

### What browsers are you seeing the problem on?

Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
looks like incorrect `append('')` usage

https://react-hook-form.com/docs/usefieldarray

![Screenshot 2024-06-15 at 8 24 04 pm](https://github.com/react-hook-form/react-hook-form/assets/10513364/aed0ef8d-58ba-492a-8e29-36fe6db4b804)

**Reply by deshiknaves** [reply]
If the field of the array is a `string` only field, then that is the correct empty value for that field. Essentially this choice of API means that you can never have an array of primitive values, only objects. I can easily replicate the functionality with reducer and `remove` from `useFieldArray` but this seems like an unnecessary limitation.

**Comment by lucasrabiec** [comment]
Same problem. It would be nice to have the possibility to provide primitive values (especially 0) to the field array.

**Comment by daniel-barhanin** [comment]
I'm having the same problem, I can't add 0 after 0, will there be a fix for it soon?
